src/stirling/source_connectors/perf_profiler/pprof/pprof.cc

```diff
@@ -0,0 +1,148 @@
+    const std::vector<std::string_view> symbols = absl::StrSplit(stack_trace_str, ";");
+
+    // Iterate over the symbols in this stack trace.
+    // Each symbol will be added to the sample as a "location" message, which in turn refers to a
+    // string in the strings table (tracked in map "strings") (more details below).
+    // Note, because of how we built the stack trace string and how the pprof profile is
+    // organized, we iterate in reverse order.
+    for (auto symbols_iter = symbols.rbegin(); symbols_iter != symbols.rend(); ++symbols_iter) {
+      // For convenience.
+      const auto& symbol = *symbols_iter;
+
+      // try_emplace() looks up an existing entry in the strings table, or creates a new entry
+      // with the key provided. Here a new entry maps from symbol to next_location_id.
+      const auto [strings_iter, inserted] = strings.try_emplace(symbol, next_location_id);
+
+      if (inserted) {
+        // New symbol: create a new location, add it to the sample, and create a new symbol.
+
+        // For convenience & clarity, store the ids in locals, then compute their next values.
+        const uint64_t location_id = next_location_id;
+        const uint64_t string_id = next_string_id;
+        ++next_location_id;
+        ++next_string_id;
+
+        // Add a new "location" to the profile.
+        // Each sample is a sequence of locations. The locations, in their sequence, represent a
+        // stack trace. Each location may include an address and a reference to a mapping (useful
+        // if symbols are not included in the profile, enables post-hoc symbolization).
+        // Lacking both address and mapping here, we skip those.
+        auto location = profile.add_location();
+        location->set_id(location_id);
+
+        // To connect a location to a symbol, we need to go through a "line" and a "function".
+
+        // Add a "line" message to the location message.
+        // In our usage, this is essentially a pointer to a function message that points to a
+        // symbol. The function message may contain more information (e.g. starting line number).
+        auto line = location->add_line();
+        line->set_function_id(location_id);
+
+        // Add a "function" to the profile (to point to the string table).
+        auto function = profile.add_function();
+        function->set_id(location_id);
+
+        // Add a reference to the string table entry in the function message.
+        function->set_name(string_id);
+
+        // Add the string to the string table in the profile.
+        profile.add_string_table(std::string(symbol));
+
+        // Add the new location-id into the sample in the profile.
+        sample->add_location_id(location_id);
+
+        // Record the fact that we have written this string into the profile. Save the
+        // location-id because it is used if we find this symbol again.
+        strings[symbol] = location_id;
```

Comment:
Isn't this covered by the `try_emplace` [here](https://github.com/pixie-io/pixie/pull/1269/files#diff-64e85fe1b081085e3ae5f47fbf8e1f768cad29707d3bb7d4267ca52af7f170c4R93)?

Reply:
Yes, good point. Fixed.